---
title: "class_exercise"
author: "Jessica Song"
date: "3/17/2023"
output: github_document
---


# Load the libraries you need
# Load functions you need "my_class_functions"
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
library(ggplot2)
source('/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/util/my_class_functions.R')
source('/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/util/_setup.R')
source('/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/util/plotting_functions.R')
setwd("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise")
```


# load in your peak files for each replicate of each protein

## import peaks

# Here I am starting to analyze my data for my proteins of interest:
H3K27me3
HDAC2
SIN3A
SIN3B
SUZ12
TBL1XR1

# First I will read in each replicate file
```{r load in peak files}

basepath <- "/scratch/Shares/rinnclass/CLASS_2023/jeso3380"
peak_path <- "CLASS_2023/CLASSES/03_Nextflow/01_my_chipseq/results/bwa/mergedLibrary/macs/broadPeak"
broadpeakfilepath <- file.path(basepath, peak_path)
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)

# printing out a table of the number of peaks in each file:

peak_numbers <- sapply(peak_list, length) %>% as.data.frame(row.names = T)
names(peak_numbers) <- c("number_of_peaks")
peak_numbers <- peak_numbers %>%
  rownames_to_column(var = "dbp") %>%
  separate(col = dbp,  into = c('dbp', 'replicate'), sep = "_")
write_csv(peak_numbers, "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/number_peaks_df.csv")

```


# Now I am going to create consensus peaks for each protein
```{r consensus peaks}
dbps <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
}))

consensus_list <- lapply(dbps, consensus_from_reduced, peak_list)
names(consensus_list) <- dbps

sapply(consensus_list, length)

number_consensus_peaks <- sapply(consensus_list, length) %>% 
  as.data.frame() %>%
  rownames_to_column( var = "dbp") %>%
  dplyr::rename(number_consensus_peaks = ".")

peak_num <- left_join(peak_numbers, number_consensus_peaks)

write_csv(peak_numbers, "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/number_peaks_df.csv")

# export consensus peaks to results folder
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/jeso3380"
consensus_path <- "/CLASS_2023/CLASSES/05_class_exercise/results/"
exportpath <- file.path(basepath, consensus_path)

for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0(exportpath, names(consensus_list)[i], "_consensus_peaks.bed") )}

```

# Now I am going to make my consensus peaks compatable with UCSC genome browser
```{r}

consensus_file_list <- list.files("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results", full.names = T, pattern = ".bed")

peaks <- lapply(consensus_file_list, read.table, col.names = c("chr", "start", "end", "name", "score", "strand"))

names(peaks) <- dbps

canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")

peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))

new_filenames <- paste0("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/00_consensus_peaks/", names(peaks), "_consensus.bed")

for(i in 1:length(peaks)) {
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

headers <- paste0("track type=bed name=", names(peaks))
headers

new_filenames <- paste0("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/UCSC_consensus_peaks/", names(peaks), ".bed")
new_filenames

# print out consensus peak files in a results/UCSC directory
for(i in 1:length(peaks)) {
  writeLines(headers[[i]], new_filenames[[i]])
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

for(i in 1:length(peaks)) {
  writeLines(headers[[i]], new_filenames[[i]])
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

```

# I am curious if my proteins are transcription factors so I will use the annotations
# in a cell paper I found and see

```{r download TF annotation data}

*** this is switched from the 01_peaks_features.Rmd that we did in class
Got gencode gene annotations before doing this step, not sure if the num_peaks_df file needs ot have that info before we do this.

basepath <- "/scratch/Shares/rinnclass/CLASS_2023/jeso3380"
peak_path <- "CLASS_2023/CLASSES/05_class_exercise/00_consensus_peaks"
consensusPeakPath <- file.path(basepath, peak_path)
setwd("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/01_peak_features")

url <- "https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx"
destination_for_url <- "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/TF_annotations.xlsx"
download.file(url, destination_for_url)

human_tfs <- readxl::read_excel("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/TF_annotations.xlsx",
                                sheet = 2, skip = 1)

names(human_tfs)[4] <- "is_tf"

***having issues with filepathing hereeeeeeee***
  
length(which(tolower(number_peaks_df.csv$dbp) %in% tolower(human_tfs$Name)))

length(which(tolower(number_peaks_df$dbp) %in% tolower(human_tfs$Name)))


human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/num_peaks_df$dbp), 1:4]


names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)
dim(num_peaks_df[is.na(num_peaks_df$tf),])

num_peaks_df <- num_peaks_df[,1:12]
write_csv(num_peaks_df, "results/num_peaks_df.csv")






basepath <- "/scratch/Shares/rinnclass/CLASS_2023/jeso3380"
peak_path <- "CLASS_2023/CLASSES/05_class_exercise/00_consensus_peaks"
consensusPeakPath <- file.path(basepath, peak_path)
setwd("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/01_peak_features")
---
consensus_peaks_files <- list.files(consensusPeakPath, 
                                             pattern = "*.bed",
                                             full.names = TRUE)

consensus_peaks <- lapply(consensus_peaks_files, rtracklayer::import)
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/00_consensus_peaks/consensus_peaks|_consensus.bed","", consensus_peaks_files)
---
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

gencode_genes <- gencode_gr[gencode_gr$type == "gene"] 
table(gencode_gr$type)
rtracklayer::export(gencode_genes, "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/gene_annotations/gencode_genes.gtf")

mrna_genes <- gencode_genes[gencode_genes$gene_type %in% "protein_coding"] 
rtracklayer::export(mrna_genes, "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/gene_annotations/mrna_genes.gtf")
table(gencode_genes$gene_type)

lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% "lncRNA"] 
rtracklayer::export(lncrna_genes, "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/gene_annotations/lncrna_genes.gtf")

mrna_lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% c("protein_coding","lncRNA")]
rtracklayer::export(mrna_lncrna_genes, "/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/gene_annotations/mrna_lncrna_genes.gtf")

lncrna_mrna_genes <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/jeso3380/CLASS_2023/CLASSES/05_class_exercise/results/gene_annotations/mrna_lncrna_genes.gtf")
lncrna_mrna_genes_df <- lncrna_mrna_genes %>% as.data.frame()


# if you leave the object name you just created in the environment
# it will print out in the knit. For example :
objectname
```




# Now I want to compare a protein with a previous analysis 
```{r}

# goto UCSC genome browser and load in a peak file for a given protein
# load in the data for the same protein from the previous analysis
# compare how your consensus peaks are similar or different to previous analyses



```


# Now I am going to determine how my peaks for each protein overlap annotations of the genome
# First I will find the overlaps between my consensus peaks with promoters of lncRNA and mRNA promoters

```{r}


# find overlaps of promoters for each protein

```

## results: 
#1) What can you determine from these overlaps?



# Now I want to compare the overlaps with lncRNA and mRNA promoters seperately 
```{r}

```
## results:
# 1) What is the difference in overlaps between mRNA and lncRNA promoters


# Now I am going to test if there is more binding over gene bodies than promoters
# I will seperate lncRNA and mRNA gene bodies to find the overlaps 

```{r}



```
## results: 
# 1) Do my proteins have more overlaps with promoters or genebodies?



# It is nice and all to find overlaps, but I am interested in how many proteins
# bind a specific promoter. I will use my handy "occurence" parameter in 
# " count peaks per feature" 

```{r}

```
## results: I find the max number of proteins on a promoter to be X


# Now I want to start plotting my results
# First I will see if there is a realtionship between peak number and total DNA covered
```{r}


ggplot ()
```

# Now I want to color my plot by wether the protein is a TF or not.
```{r}


ggplot
```

# I want to make a histogram of the number of peaks for each of my proteins

```{r}



hist
```


# Now I want to facet this by the type of DNA binding domain my protein has.
```{r}

```


# Cool now I am ready to send my result to my collaborator as a
# Knitted document
