---
title: "05_Genomic_Features_in_Granges"
author: "JR"
date: "7/21/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
# library(Gviz)
library(IRanges)

```

Accompanying Youtube video of this class:
https://youtu.be/l1AcyUdNCDc

# Let's go explore the human geneome annotations from GENCODE.
We will do this by importing a GTF file of these annotations.
A GTF file is similar in a way to a .bed file with start stop of features
and other associated metadata that is described here:

https://uswest.ensembl.org/info/website/upload/gff.html

(1) Need GEncode GTF of annotations for this class. # Let's go get all the features of the human genome ! I have downloaded this from gencode and stored for everyone below. 

If not in class please download Gencode annotations gtf here : https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/


## Ok let's start exploring the features of the human genome!
```{r loading in GENCODE GRanges annotation file}

# setting class path to Gencode annotations GTF
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes"

# we use rtracklayer::import to turn the annotations into start and stops (Iranges) w/ chr (Granges)This takes ~5min

gencode_gr <- rtracklayer::import(file.path(basepath, "gencode.v32.annotation.gtf"))

# let's take a look @metadata to see what we have in here

```

Now let's look at the gene_id indexable part of this GRanges
# seqnames
```{r seqnames}

?seqnames
  
seqnames(gencode_gr) %>% 
  head()

# we can see that "levels" are the chromosome names and are a meta-level
```

let's get more info with table to summary using tidy verse %>%
# summary / table a GRanges object
```{R summary/table GRanges}

table(seqnames(gencode_gr)) %>%
  head()
# This is telling us how many features are associated with each chromosome.

table(seqnames(gencode_gr)) %>% 
  summary()
# this tells us how many features there are in the genome.

# Let's find out how many features the mitochondrial chr has (25th level)
table(seqnames(gencode_gr))[25] 

# we see this is the mitochondrial chromosome and has 143 elements

```

# Indexing Granges (@elementMetadata$INDEXING)
```{R indexing GRanges}

# we see a lot of good information in @elementMetadata -- let's index it.

# let's see how many 'gene_id' values there are by indexing
table(gencode_gr@elementMetadata$gene_id) %>% 
  head()
# This tell us how many features are associated with each gene id.

# Let's see what we get with gene_name
table(gencode_gr@elementMetadata$gene_name) %>% 
  head()

# Typical gene symbols 
table(gencode_gr@elementMetadata$transcript_name) %>%
  head()

# transcript_id
## for a specific version of a gene

# it's basically memorization of what is a gene name or gene id 

# if you wanted to look at this more expansively (instead of just head lines)
# you can make it into an object

gencode_info <- table(gencode_gr@elementMetadata$gene_name) %>% 
 as.data.frame()
# you can pipe one object into another object

# cool now you can look for any feature for your favorite genes
```


Let's analyze some of the features of our genome!
# Making a data frame of items in GRanges object
```{r GRanges indexing to data.frame}

# above we tested different indexes and made a dataframe of one feature
# now let's do this multiple times at once!

gencode_gr_length <- data.frame("gene_id" = gencode_gr@elementMetadata$gene_id,
                           "gene_name" = gencode_gr@elementMetadata$gene_name,
                           "gene_type" = gencode_gr@elementMetadata$gene_type,
                           "type" = gencode_gr@elementMetadata$type,
                           "start" = gencode_gr@ranges@start,
                           "width" = gencode_gr@ranges@width,
                           "chr" = gencode_gr@seqnames)
                           
```

Ok so now we have a new data frame with the information we want from gencode_gr
Note we see that we can index the RANGES for each gene with @ranges "level"


# Let's explore features using "filter()"

```{R mRNA gene width}

# here we are filtering our dataframe above using the column names (e.g., gene_type)
# parameters for filter(data.frame, col-name1 == " character" )
# We will filter to just "protein_coding" genes

mrna_df <- filter(gencode_gr_length, gene_type == "protein_coding", type == "gene")

# we can easily plot a histogram of mRNA gene widths.
hist(log10(mrna_df$width), breaks = 60)

# same for lncRNAs
lncrna_df <- filter(gencode_gr_length, gene_type == "lncRNA", type == "gene")

# plotting lncRNA
hist(log10(lncrna_df$width), breaks = 60)

# Let's see what the average gene size is. Note we can no longer use width in a DF
mean(lncrna_df$width)
mean(mrna_df$width)

# this all points to mRNAs being 2x longer than lncRNAs.
# Note this is for genebody length -- we will do more with exons below.
```


# How many types of "gene types" are there?

Here we will index into gencode_gr to $type
this has all the info of types of genes included.
```{R gene types}

# indexing into gene typse 
table(gencode_gr@elementMetadata$type)

## cool this tell us a lot about the genome with one line!

# Moving forward we will use the "$type" "genes"
```

# Finding genome feature lengths
let's find the longest and shortest genes in the genome!
But first let's see how to find the lengths of genome features.
```{R annotation lengths} 

# Let's see the range of gene sizes
summary(width(gencode_gr))

```

interesting some genes are just 1 nt long!
let's look into that some more

# what annotations have width of 1nt?
```{R annotation widths of 1 ?}
# we will use indexing and setting a value of 1 and making DF
width_1_items <- gencode_gr[width(gencode_gr) == 1] %>%
  as.data.frame()

# let's look
table(width_1_items$type)
# so weird that a CDS can be 1nt or an exon.

# looking for weird exons (also CDS and codons don't make sense to be = 1??)
weird_exons <- width_1_items %>% 
  filter(type == "exon")

```

huh 33 exons are 1 nt -- annotations are never perfect :)
The one thing that should be 1nt long is a TSS.
To get the TSS we need to use the promoters function:

# defining gene TSS 
```{R annotation widths of 1 ?}

# The promoters function defines the TSS and input bases up and downstream.
# Parameters are some gene annotation of interest: let's use "gene"
tss <- promoters(gencode_gr[gencode_gr$type == "gene"], 
                 upstream = 0, downstream = 0) %>%
  as.data.frame()

# we see 60,609 TSS. Let's see if that is the same as # of genes
summary(gencode_gr[gencode_gr$type == "gene"])
# cool the same

```

# What is the shortest gene?
Here we will bring together our indexing and filter skills
```{shortest gene in human genome}
# Let's turn gencode_gr into a data.frame and filter to only genes.
gencode_df <- gencode_gr %>%
  as.data.frame() %>%
  filter(type == "gene")

# Let's find the smallest genes using "arrange" function
gencode_df %>% 
  arrange(gencode_df$width) %>%
  head()

# Voila here is the shortest gene (8nt long ??): TRDD1
# seqnames     start       end width strand source type score phase           gene_id gene_type gene_name level    hgnc_id          havana_gene
# 1    chr14  22438547  22438554     8      + HAVANA gene    NA    NA ENSG00000223997.1 TR_D_gene     TRDD1     2 HGNC:12254 OTTHUMG00000170844.2
```

# now the longest gene using -arrange
```{longest  gene in human genome}

# Now let's find the Largest genes using arrange
gencode_df %>%
  arrange(-width) %>%
  head()

#   seqnames     start       end   width strand source type score phase            gene_id      gene_type gene_name level    hgnc_id
# 1    chr16   5239802   7713340 2473539      + HAVANA gene    NA    NA ENSG00000078328.21 protein_coding    RBFOX1     1 HGNC:18222

# We can also store the longest gene as an object with max function
longest_gene <- gencode_df %>%
  filter(width == max(width))

# let's take a look at what information we have for this gene:
longest_gene

```

Now let's find the total amount of the genome covered by exons.
We start by defining total_exon_length of all exons in genome.

# we need to use reduce to collapse all the exons that overlap into the longest exon
# Exon 1.1  _____________
# Exon 1.2      _____________
# reduce    _________________
# intersect     _________


Ok let's find out what % of genome is exons!

# What percetnage of the genome (3Billion) are exons?
```{R % of genome that is exons}
# first we will make an object of exons and reduce them as described above
total_exon_length <- gencode_gr[gencode_gr$type == "exon"] %>%
  # now reduce will take the union of overlapping exons 
  GenomicRanges::reduce() %>% 
  width() %>%
  sum()

# We now have the total bases covered by exons in the human genome. We also know
# the human genome is 3.9B base pairs so let's find the percentage by division:

total_exon_length/3.2e9
# Boom we can see the human genome is about 4% exons!

# if we want to plot the range of exon sizes we simply remove the sum function and plot:
exon_lengths <- gencode_gr[gencode_gr$type == "exon"] %>%
  # now reduce will take the union of overlapping exons 
  GenomicRanges::reduce() 

# plotting
hist(width(exon_lengths), breaks = 100, xlim = c(0,10000))
# we see most exons are under 2kb what is the average?

avergage_exon_length <- mean(width(exon_lengths))
# Cool the average exon length is 444nt !
```

# what percentage of the genome is comprised of gene bodies?
```{R % of genome that is Gene bodies}

# To do this we simply change index in $type to == gene
reduced_gene_bodies <- gencode_gr[gencode_gr$type == "gene"] %>%
  GenomicRanges::reduce() %>%
  width() %>%
  sum()

reduced_gene_bodies/3.2e9
# Woah gene bodies take up 50% of genome !!

# lets' see the range of sizes (take out width adn sum from above)
gene_body_lengths <- gencode_gr[gencode_gr$type == "gene"] %>%
  GenomicRanges::reduce()

# plotting
hist(width(gene_body_lengths), breaks = 100, xlim = c(0,500000))

# we see most are less than 200Kb long (we previous saw longest was 2M !)

# what is the average length of a gene body?

mean(width(gene_body_lengths))

# answer is 38Kb !
```

####################
Excercise
###################

Determine:

1) How many CDS (Coding sequences there are in the genome)
2) What is the longest and shortest CDS?
3) What is the distribution of CDS lengths?


